sdks/python/apache_beam/pipeline.py

```diff
@@ -206,6 +205,22 @@ def _replace_if_needed(self, transform_node):
             # When current transform is a primitive, we set the producer here.
             new_output.producer = transform_node
 
+          # We only support replacing transforms with a single output with
+          # another transform that produces a single output.
+          # TODO: Support replacing PTransforms with multiple outputs.
+          if len(transform_node.outputs) > 1:
+            raise NotImplementedError(
+                'PTransform overriding is only supported for PTransforms that '
+                'has a single output. Tried to replace %r that has %d outputs.'
+                , transform_node, len(transform_node.outputs))
+
+          if type(new_output) is tuple:
```

Comment:
This is not sufficient, lists or dicts (or other types) could be returned. Only allow a single PCollection (and likewise that the previous was None -> PCollection)?